param-plugin/params.cc

```diff
@@ -0,0 +1,76 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C) 2012  Clifford Wolf <clifford@clifford.at>
+ *  Copyright (C) 2020  The Symbiflow Authors
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+#include <cassert>
+#include "kernel/register.h"
+#include "kernel/rtlil.h"
+#include "kernel/log.h"
+
+USING_YOSYS_NAMESPACE
+
+PRIVATE_NAMESPACE_BEGIN
+
+
+void register_in_tcl_interpreter(const std::string& command) {
+	Tcl_Interp* interp = yosys_get_tcl_interp();
+	std::string tcl_script = stringf("proc %s args { return [yosys %s {*}$args] }", command.c_str(), command.c_str());
+	Tcl_Eval(interp, tcl_script.c_str());
+}
+
+struct GetParam : public Pass {
+	GetParam() : Pass("get_param", "Convert string params to floats") {
+		register_in_tcl_interpreter(pass_name);
+	}
+
+	void help() override
+	{
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("   convert_params_to_real \n");
+		log("\n");
+		log("Process the real parameters as floats instead of strings \n");
+		log("\n");
+	}
+
+	void execute(std::vector<std::string> args, RTLIL::Design* design) override
+	{
+		if (args.size() == 1) {
+			log_error("Incorrect number of arguments");
+		}
+		std::string param(args.at(1));
+		std::string value;
+		extra_args(args, 2, design);
+
+		for (auto module : design->selected_modules()) {
+			for (auto cell : module->selected_cells()) {
+				auto params = cell->parameters;
+				auto it = params.find(RTLIL::IdString(RTLIL::escape_id(param)));
+				if (it != params.end()) {
+					value = it->second.decode_string();
+				}
+			}
+		}
+		Tcl_Interp *interp = yosys_get_tcl_interp();
+		Tcl_SetResult(interp, const_cast<char*>(value.c_str()), NULL);
```

Comment:
This is not legal C++, because the malloc'd array in value leaves scope.

Reply:
From the documentation on `Tcl_SetResult`:

```
Tcl_SetResult arranges for result to be the result for the current Tcl command in interp, replacing any existing result. The freeProc argument specifies how to manage the storage for the result argument; it is discussed in the section THE TCL_FREEPROC ARGUMENT TO TCL_SETRESULT below. If result is NULL, then freeProc is ignored and Tcl_SetResult re-initializes interp's result to point to an empty string.
```

You need to provide a `freeProc` per the documentation to free the string once the interperator is done with it.

Reply:
Reading further, it looks like you need to allocation a new Tcl memory space using https://www.tcl.tk/man/tcl8.5/TclLib/Alloc.htm and set freeProc = `TCL_DYNAMIC`.

Reply:
Thanks for the info. I remembered there is something to do with memory to make the Tcl function in C++ work, hence this is still WIP.

Reply:
Fixed.